File: algorithms/[992]K 个不同整数的子数组/solution.ts
// topics = ["滑动窗口", "哈希表"]

function subarraysWithKDistinct(A: number[], K: number): number {
  const n = A.length;

  const atMostK = (x: number): number => {
    let i = 0;
    let j = 0;
    let res = 0;

    const map = new Map<number, number>();

    while (j < n) {
      if (map.has(A[j])) {
        map.set(A[j], map.get(A[j])! + 1);
      } else {
        map.set(A[j], 1);
      }

      while (map.size > x) {
        map.set(A[i], map.get(A[i])! - 1);
        if (map.get(A[i]) == 0) {
          map.delete(A[i]);
        }
        i += 1;
      }

      j += 1;
      res += j - i;
    }

    return res;
  };

  // 恰好包含 K 种不同整数 = 最多包含 k 种不同整数 - 最多包含 k - 1 种不同整数
  return atMostK(K) - atMostK(K - 1);
}

File: algorithms/[992]K 个不同整数的子数组/solution.py
# topics = ["滑动窗口", "哈希表"]

from collections import Counter
from typing import List, Dict


class Solution:
    def subarraysWithKDistinct(self, A: List[int], K: int) -> int:
        """
        近似题 904.水果成篮
        恰好包含 K 种不同整数 = 最多包含 k 种不同整数 - 最多包含 k - 1 种不同整数
        """
        return self.atMostK(A, K) - self.atMostK(A, K - 1)

    def atMostK(self, A: List[int], K: int) -> int:
        n = len(A)
        i = j = 0
        res = 0

        c: Dict[int, int] = Counter()

        while j < n:
            c[A[j]] += 1

            while len(c) > K:
                c[A[i]] -= 1
                if c[A[i]] == 0:
                    c.pop(A[i])
                i += 1

            j += 1
            # 新增的子数组个数（以 j 为右边界）为 j - i
            res += j - i

        return res
